workers/unity/Packages/com.improbable.gdk.core/GameObjectRepresentation/GameObjectDispatcherSystem.cs

```diff
@@ -0,0 +1,161 @@
+            {
+                gameObjectComponentDispatcher.ComponentAddedComponentGroup =
+                    GetComponentGroup(gameObjectComponentDispatcher.ComponentAddedComponentTypes);
+                gameObjectComponentDispatcher.ComponentRemovedComponentGroup =
+                    GetComponentGroup(gameObjectComponentDispatcher.ComponentRemovedComponentTypes);
+                gameObjectComponentDispatcher.AuthoritiesChangedComponentGroup =
+                    GetComponentGroup(gameObjectComponentDispatcher.AuthoritiesChangedComponentTypes);
+                if (gameObjectComponentDispatcher.ComponentsUpdatedComponentTypes.Length > 0)
+                {
+                    gameObjectComponentDispatcher.ComponentsUpdatedComponentGroup =
+                        GetComponentGroup(gameObjectComponentDispatcher.ComponentsUpdatedComponentTypes);
+                }
+
+                if (gameObjectComponentDispatcher.EventsReceivedComponentTypeArrays.Length > 0)
+                {
+                    gameObjectComponentDispatcher.EventsReceivedComponentGroups =
+                        new ComponentGroup[gameObjectComponentDispatcher.EventsReceivedComponentTypeArrays.Length];
+                    for (var i = 0; i < gameObjectComponentDispatcher.EventsReceivedComponentTypeArrays.Length; i++)
+                    {
+                        gameObjectComponentDispatcher.EventsReceivedComponentGroups[i] =
+                            GetComponentGroup(gameObjectComponentDispatcher.EventsReceivedComponentTypeArrays[i]);
+                    }
+                }
+
+                if (gameObjectComponentDispatcher.CommandRequestsComponentTypeArrays.Length > 0)
+                {
+                    gameObjectComponentDispatcher.CommandRequestsComponentGroups =
+                        new ComponentGroup[gameObjectComponentDispatcher.CommandRequestsComponentTypeArrays.Length];
+                    for (var i = 0; i < gameObjectComponentDispatcher.CommandRequestsComponentTypeArrays.Length; i++)
+                    {
+                        gameObjectComponentDispatcher.CommandRequestsComponentGroups[i] =
+                            GetComponentGroup(gameObjectComponentDispatcher.CommandRequestsComponentTypeArrays[i]);
+                    }
+                }
+            }
+        }
+
+        protected override void OnUpdate()
+        {
+            foreach (var gameObjectComponentDispatcher in GameObjectComponentDispatchers)
+            {
+                gameObjectComponentDispatcher.InvokeOnAddComponentLifecycleCallbacks(this);
+                gameObjectComponentDispatcher.InvokeOnRemoveComponentLifecycleCallbacks(this);
+                gameObjectComponentDispatcher.InvokeOnAuthorityChangeLifecycleCallbacks(this);
+            }
+
+            foreach (var spatialOSBehaviourManager in spatialOSBehaviourManagers)
+            {
+                    spatialOSBehaviourManager.EnableSpatialOSBehaviours();
```

Comment:
If the component is removed I think we don't actually want to receive callbacks. That will create a null ref if they try to read the component during the update.

Reply:
Similarly this will cause callbacks on writers to give callbacks for your previous frame's updates, which might trigger sending another update, thinking the authority is still there.

Reply:
I'm not sure if I'm following. This is not the place in the code where user updates would be triggered. If a component is removed or authority is lost during a frame, the assumption is that the GDK core would not give you the respective reactive component tags.

Reply:
The comment was about the ordering in which you enable update and disable monobehaviour callbacks. If a component has been removed you probably want to disable the stuff before giving callbacks on it. If you have two readers for different types on a single monobehaviour, then one could try to read the data from the other during a callback. If one has been removed and the other has a callback the current ordering will cause that to null ref. Which breaks the design that an enabled monobehaviour is good to go.

To fix this you want to disable things before giving callbacks.

However you also want to make sure you aren't doing things like trigger update callbacks on writers. So you can't just enable everything before the start and naively proceed. You probably want to enable behaviours with writers without invoking their callbacks for updates on that component.

So the overall order likely needs to be
Disable stuff
Enable stuff without writers
Deal with auth lost and auth loss immanent
Deal with component updates
Deal with auth gained
Enable stuff with writers

Reply:
This is a very good point and the more I look at the current code the more flawed the execution order looks. Agreed that we need to disable everything first. I still don't understand why we need to distinguish our behaviour for readers and writers though. 
- An authority gained callback can be triggered for both readers and writers after the enabling phase.
- Authority loss iminent notifications can be received by both readers and writers in the same way.
- Re: Component updates: As mentioned in my other post, this matter should be solved on the GDK Core level.
- Authority loss callbacks are "automatically" not fired on writers assuming that they get disabled earlier in the frame.

Reply:
Created https://improbableio.atlassian.net/browse/UTY-830

Reply:
It would be confusing to receive auth added callbacks on writers. As a user I'd expect that a writer is only enabled when I have authority. If I saw I could get an auth callback after OnEnable I'd be confused about when I'm allowed to send updates. On a writer enabled behaviour you might be getting auth callbacks for auth loss imminent. If you got a callback with something else in it then users need to write code to call that too and would likely be considered a bug.

You're right auth loss imminent won't break things, so my execution order is probably wrong. As an aside we might need to be more careful about it as we add things (e.g. readonlyers or abouttoloseauthwriters). But we don't have those yet.

Unfortunately you can't assume that about the core. Updates do have to be handled more carefully :(.

Yes, the disabling of stuff earlier will solve that.